Restore some functionality of the sphinx directive.

Carreau opened the pull request:
See #11362

The issue is in 2 part, before IPython 7.0 the input splitter was
state full, this was (in part) due to readline.

The second part is that because of this, we had to be a bit adressive of
what was considered complete code (it had to have 2 new line). This is
now not required anymore as we can submit stuff as a whole.

I hope that this fixes that. I have another fix in mind that count (and
reset) the number of consecutive blank line, but that will be more
complicated end code.

Carreau committed:
Restore some functionality of the sphinx directive.

See #11362

The issue is in 2 part, before IPython 7.0 the input splitter was
state full, this was (in part) due to readline.

The second part is that because of this, we had to be a bit adressive of
what was considered complete code (it had to have 2 new line). This is
now not required anymore as we can submit stuff as a whole.

I hope that this fixes that. I have another fix in mind that count (and
reset) the number of consecutive blank line, but that will be more
complicated end code.

TomAugspurger commented:
This fixed things for the small test in #11362 

I'm giving it a try on pandas docs.

TomAugspurger commented:
It seems like exceptions fail the build. With the same setup from #11362, 

```python
# index.rst
Welcome to test's documentation!
================================

.. ipython:: python

   def f(x):
       x = 2
       raise ValueError('bad')
       return x

   f(2)

.. ipython:: python

   print('hi')


Testing
```

then `make html` results in 

```
Running Sphinx v1.8.1
making output directory...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: 1 added, 0 changed, 0 removed
reading sources... [100%] index

>>>-------------------------------------------------------------------------
Exception in /Users/taugspurger/sandbox/sphinx-test/index.rst at block ending on line 12
Specify :okexcept: as an option in the ipython:: block to suppress this message
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-c510dc86724b> in <module>
----> 1 f(2)

<ipython-input-1-fe8b2c77be70> in f(x)
      1 def f(x):
      2     x = 2
----> 3     raise ValueError('bad')
      4     return x

ValueError: bad
<<<-------------------------------------------------------------------------


Exception occurred:
  File "/Users/taugspurger/sandbox/IPython/IPython/sphinxext/ipython_directive.py", line 562, in process_input
    raise RuntimeError('Non Expected exception in `{}` line {}'.format(filename, lineno))
RuntimeError: Non Expected exception in `/Users/taugspurger/sandbox/sphinx-test/index.rst` line 12
The full traceback has been saved in /var/folders/hz/f43khqfn7b1b1g8z_z6y3bsw0000gp/T/sphinx-err-6psc98eb.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [html] Error 2
```

I I include `okexcept` as an argument to the directive, things do finish. IIRC, the behavior on 6.x was to print a warning to stdout with a message telling you to use `okexcept`.

Carreau commented:
> I I include `okexcept` as an argument to the directive, things do finish. IIRC, the behavior on 6.x was to print a warning to stdout with a message telling you to use `okexcept`.

Yes but that was changed explicitley in https://github.com/ipython/ipython/pull/11321 (should maybe have been better documented) to actually start to catch regression in documentation.... 

I can try to make it configurable though. One other discussion was to make `okexcept` fail if there is actually no exception...

Carreau committed:
Add an option (`ipython_warning_is_error`) to not stop on error.

The behavior pre-6.5 was to keep on going even if unexpected exceptions
or warnings were shown, on 7.x the default is to abort the build.

For compat reasons (and convenience), you can now set back the behavior
to the original one to just log to stderr and move on.

Carreau commented:
You can now set `ipython_strict_fail = False` in your `config.py` to get back previous behavior. 

if you have a better name...

TomAugspurger commented:
Thanks, that name seems fine, but if you want to match Sphinx I think it would be `ipython_warning_is_error` to match `warning-is-error`.

Carreau committed:
Catch Syntax error as exceptions, get doc from history

Carreau commented:
> `ipython_warning_is_error` to match `warning-is-error`

Done, 

SyntaxError is now detected as a error as well, I restored the old documentation from 0.13... that may need some polish.

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+
+.. note::
+
+   This has been salvadged from history, so information may be approximate or
```
I'd just say something like "The IPython Sphinx Directive is in 'beta' and currently under active development. Improvements to the code or documentation are welcome!"

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+doctest the output.  The inputs are fed to an embedded ipython
+session and the outputs from the ipython session are inserted into
+your doc.  If the output in your doc and in the ipython session don't
+match on a doctest assertion, an error will be
```
error will be...

Carreau replied:
Yes, feel free to make any replacement you feel is better.

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+   In [1]: x = 'hello world'
+
+   # this will raise an error if the ipython output is different
+   @doctest
```
is @doctest explicitly documented somewhere? or only in reference here?

LucianaMarques replied:
Could someone please tell me more about @doctest? Where is it documented?

LucianaMarques replied:
Ok, now I see there's a definition at the bottom of the file. But some more info on it would be nice, I find it a little confusing.

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+  SyntaxError: invalid syntax
+
+
+The embedded interpreter supports some limited markup.  For example,
```
I'd split each of these sub-sections by a header. E.g.

```
Persisting the Python session across IPython directive blocks
=============================================================

...

Adding documentation tests to your IPython directive
====================================================

...

Multi-line input
================

etc

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+
+You can do doctesting on multi-line output as well.  Just be careful
+when using non-deterministic inputs like random numbers in the ipython
+directive, because your inputs are ruin through a live interpreter, so
```
*run

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+You can do doctesting on multi-line output as well.  Just be careful
+when using non-deterministic inputs like random numbers in the ipython
+directive, because your inputs are ruin through a live interpreter, so
+if you are doctesting random output you will get an error.  Here we
```
I think that the "random output" comparison step can be a bit clearer. Something like:

```
Testing directive outputs
=========================

The IPython Sphinx Directive makes it possible to test the outputs that you provide with your code. To do this, decorate the contents in your directive block with one of the following:

  * list directives here

If an IPython doctest decorator is found, it will take these steps when your documentation is built:

1. Run the *input* lines in your IPython directive block against the current Python kernel (remember that the session persists across IPython directive blocks).
2. Compare the *output* of this with the output text that you've put in the IPython directive block (what comes after `Out[NN]`)
3. If there is a difference, the directive will raise an error and your documentation build will fial.

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+
+   In [152]: title('normal distribution')
+
+   @savefig hist_with_text.png width=4in
```
maybe this should be called `display_fig` or `show_fig` instead of savefig? From a user's standpoint I imagine they'd think "I want to show this figure" not "I want to save this figure"

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+
+.. ipython:: python
+
+   @savefig plot_simple_python.png width=4in
```
will this plot a figure (because of `@savefig`) or will it plot nothing (because of the semicolon)?

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+   ioff()
+   ion()
+
+Multi-line input is handled.
```
People could confuse this with multi-line input in the ipython section above. Maybe call this "splitting Python statements across lines"?

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+           works'
+   print(line.split('&'))
+
+Functions definitions are correctly parsed
```
I'd have a section called "syntax highlighting" and just step through several major use-cases there. E.g.:

```
Syntax highlighting
===================

The Sphinx IPython directive can highlight most Python structures and patterns. This section shows some of the most common ones.

Multi-line statements
---------------------
EXAMPLE CODE

Function definitions
--------------------
EXAMPLE CODE

etc
```

LucianaMarques replied:
Hi @choldgraf , I don't know if I understood quite well what you meant here. 

Do you mean creating another section with a whole set example code blocks or rename an existing section for Syntax Highlighting and use the code that has already been written?

Thank you

Carreau replied:
I think chris was suggesting refactoring and being clearer with heading/subheading. 

You can see the rendered version [there](https://ipython.readthedocs.io/en/latest/development/ipython_directive.html), There is a lot of duplicate information and organisational issues.

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+    Also, can be applied to the entire ``.. ipython`` block as a
+    directive option with ``:verbatim:``.
+
+@savefig OUTFILE [IMAGE_OPTIONS]
```
Ah ok - so the decorators are here. In that case you should always link to this section whenever you mention a decorator above. Something like `for more information on decorator usage, see :ref:`directive_decorators` and then add the label `.. _directive_decorators:` to this section

LucianaMarques replied:
Is this supposed to create some kind of a link? I didn't understand much how am I supposed to add this reference. But good idea, this certainly helps.

Carreau replied:
No, that is just an example on how to use `savefig`. Do what you can, even partially. If it's annoying or confusing send what you have and feel free to move on to something more interesting.

LucianaMarques replied:
Hey Carreau, I think I had this doubt because I was not familiar with reStructureText until today. Thank you for your reply!

choldgraf commented on docs/source/development/ipython_directive.rst:
```diff
@@ -0,0 +1,435 @@
+
+Configuration Options
+=====================
+
```
Specify that these are the configurations you'd provide underneath the directive with `:option:`

choldgraf reviewed (commented):
In general it looks quite good - I have a few thoughts for clarifying/structuring the documentation!

Carreau committed:
fix minimal issues

Carreau committed:
install matplotlib to build docs

Carreau commented:
Thanks @choldgraf for the feedback, I've open #11410 as a easy first issue to follow up on. 

I'm going to get that in as it fixes some functionalities.

Carreau commented:
No problem, restructured text can be complicated at time.

On Sun, Oct 21, 2018, 10:49 Luciana da Costa Marques <
notifications@github.com> wrote:

> *@LucianaMarques* commented on this pull request.
> ------------------------------
>
> In docs/source/development/ipython_directive.rst
> <https://github.com/ipython/ipython/pull/11402#discussion_r226871658>:
>
> > +
> +@suppress
> +
> +    execute the ipython input block, but suppress the input and output
> +    block from the rendered output.  Also, can be applied to the entire
> +    ``.. ipython`` block as a directive option with ``:suppress:``.
> +
> +@verbatim
> +
> +    insert the input and output block in verbatim, but auto-increment
> +    the line numbers. Internally, the interpreter will be fed an empty
> +    string, so it is a no-op that keeps line numbering consistent.
> +    Also, can be applied to the entire ``.. ipython`` block as a
> +    directive option with ``:verbatim:``.
> +
> +@savefig OUTFILE [IMAGE_OPTIONS]
>
> Hey Carreau, I think I had this doubt because I was not familiar with
> reStructureText until today. Thank you for your reply!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ipython/ipython/pull/11402#discussion_r226871658>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAUez1fiY8gzr6ON8z8Rv6StIRy3XyYRks5unLONgaJpZM4XdZXN>
> .
>
